### PR TITLE
Feature: Allow supervisors to revert thesis to previous state

### LINKED ISSUE #949: Bug: Can't revert student state from writing to proposal
When you accidentally set a student to writing, you can change the date for the writing state but it doesn't fix the state.

### PR DESCRIPTION
**Pull request by krusche**
## Summary

- Adds a **Revert to Previous State** button in the thesis Configuration section so supervisors and examiners can undo an accidental forward state transition (e.g. clicking *Accept Proposal* on the wrong thesis, or needing to reopen a `FINISHED` thesis because a required document was never uploaded).
- New `POST /v2/theses/{thesisId}/revert-state` endpoint, gated by supervisor access, that deletes the latest `thesis_state_changes` entry and sets the thesis state to the previous one. When reverting from `FINISHED` or `DROPPED_OUT`, the student group is re-added so the student regains access.
- Side data (proposal approval, assessment, final grade) is preserved so the supervisor can re-advance without re-entering values.

Closes #949.

## Why

The thesis lifecycle was forward-only. The Configuration section let supervisors edit state-change timestamps but never the current state itself, which is exactly the misbehaviour reported in the issue ("changing the date of `WRITING` does not move the thesis out of `WRITING`").

## Design

Full design doc committed at `docs/superpowers/specs/2026-05-14-revert-thesis-state-design.md`.

## Test plan

- [x] `./gradlew test --tests "de.tum.cit.aet.thesis.service.ThesisServiceTest"` — all 14 tests pass, including 4 new ones (single-state guard, basic revert, revert-from-FINISHED restores group, revert-from-DROPPED_OUT restores group)
- [x] `./gradlew test --tests "de.tum.cit.aet.thesis.controller.ThesisControllerTest"` — all 43 tests pass, including 3 new ones (happy path, bad-request on single state, 403 for non-supervisor)
- [x] `./gradlew spotlessCheck` — clean
- [x] `pnpm exec tsc --noEmit` in `client/` — clean
- [x] `pnpm exec eslint` on changed file — no new errors
- [ ] Manual smoke test: accept a proposal, then click *Revert to Previous State* and confirm the state badge returns to `PROPOSAL` and the *Accept Proposal* button reappears
- [ ] Manual smoke test: close a thesis (`DROPPED_OUT`), then revert and verify the student regains access

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * New thesis state reversion capability allows supervisors and examiners to revert a thesis to its previous lifecycle state with a confirmation flow integrated in the thesis configuration area.

* **Tests**
  * Added test coverage for thesis state reversion, including authorization checks and edge-case handling.

* **Documentation**
  * Added feature specification documenting thesis state reversion behavior, authorization requirements, and constraints.

[Review Change Stack](https://app.coderabbit.ai/change-stack/ls1intum/thesis-management/pull/1042)

### CODE REVIEW

File: server/src/main/java/de/tum/cit/aet/thesis/service/ThesisService.java [hunk 1]
```diff
@@ -842,6 +842,56 @@ public Thesis completeThesis(Thesis thesis) {
...
+	 * in {@code FINISHED} or {@code DROPPED_OUT}, the student group is restored for each
+	 * student, mirroring the inverse of {@link #completeThesis(Thesis)} / {@link #closeThesis(Thesis)}.
+	 *
+	 * @param thesis the thesis to revert
+	 * @return the updated thesis
+	 * @throws ResourceInvalidParametersException if the thesis is anonymized or has no previous state
+	 */
+	// TODO: we should avoid using @Transactional because it can lead to performance issue and concurrency problems
+	@Transactional
+	public Thesis revertToPreviousState(Thesis thesis) {
```
[the remark this answers is not in the document]
Comment (krusche) [reply server/src/main/java/de/tum/cit/aet/thesis/service/ThesisService.java]: Addressed in e4a8d5fd: removed the `@Transactional` annotation and reordered operations (thesis row save first, then state-change row delete, then student-group restore) so partial failures are retry-safe per the project guideline.

File: server/src/main/java/de/tum/cit/aet/thesis/service/ThesisService.java [hunk 2]
```diff
@@ -842,6 +842,56 @@ public Thesis completeThesis(Thesis thesis) {
...
+			throw new ResourceInvalidParametersException("Thesis has no previous state to revert to");
+		}
+
+		ThesisStateChange currentChange = orderedStates.get(0);
+		ThesisState previousState = orderedStates.get(1).getId().getState();
+		ThesisState revertedFrom = currentChange.getId().getState();
+
+		thesisStateChangeRepository.deleteById(currentChange.getId());
+		Set<ThesisStateChange> remaining = new HashSet<>(thesis.getStates());
+		remaining.remove(currentChange);
```
[the remark this answers is not in the document]
Comment (krusche) [reply server/src/main/java/de/tum/cit/aet/thesis/service/ThesisService.java]: Fixed in e4a8d5fd: switched to `removeIf(sc -> sc.getId().equals(currentChangeId))` so removal is id-based and not affected by `ThesisStateChange` lacking custom `equals`/`hashCode`.

File: server/src/main/java/de/tum/cit/aet/thesis/service/ThesisService.java [hunk 3]
```diff
@@ -842,6 +842,56 @@ public Thesis completeThesis(Thesis thesis) {
...
+	 * <p>Side data (proposal approval, assessment entities, final grade) is intentionally
+	 * preserved so the supervisor can re-advance without re-entering it. If the thesis was
+	 * in {@code FINISHED} or {@code DROPPED_OUT}, the student group is restored for each
+	 * student, mirroring the inverse of {@link #completeThesis(Thesis)} / {@link #closeThesis(Thesis)}.
+	 *
+	 * @param thesis the thesis to revert
+	 * @return the updated thesis
+	 * @throws ResourceInvalidParametersException if the thesis is anonymized or has no previous state
+	 */
+	// TODO: we should avoid using @Transactional because it can lead to performance issue and concurrency problems
```
[the remark this answers is not in the document]
Comment (krusche) [reply server/src/main/java/de/tum/cit/aet/thesis/service/ThesisService.java]: Followed the project guideline in CLAUDE.md ("Do not use @Transactional on service methods"): removed the annotation and the TODO comment in e4a8d5fd. Operations are now ordered so a partial failure is recoverable via retry — the row save happens first (user-visible), then the history-row delete, then group restore.

**Comment by krusche** [comment]
@Claudia-Anthropica review

**Comment by Claudia-Anthropica** [comment]
@krusche on it! Taking a look now :eyes:
